internal/controllers/rest_controller.go

```diff
@@ -4,52 +4,110 @@ import (
+	credentials   []handlers.CredentialsHandler
 	logger        logging.Logger
 }
 
 // NewRESTController initializes rest controller.
 func NewRESTController(
 	logger logging.Logger,
-	configuration RESTControllerConfiguration,
-	oidcService *services.OIDCService,
+	configuration configurations.RESTControllerConfiguration,
+	oidcService services.OIDCServiceInterface,
+	credentials []handlers.CredentialsHandler,
 ) RESTController {
 	return RESTController{
 		configuration: configuration,
 		oidcService:   oidcService,
+		credentials:   credentials,
 		logger:        logger,
 	}
 }
 
-// Boot initialize creating some routes.
+// Boot initializes creating some routes.
 func (rc RESTController) Boot(s *server.Server) {
-	s.Router.GET(rc.configuration.AuthServiceURL, gin.WrapF(rc.RedirectToOIDCProvider))
+	s.Router.GET(rc.configuration.AuthServiceURL, gin.WrapF(rc.AuthenticationSession))
 	s.Router.GET(rc.configuration.OIDCCallbackURL, gin.WrapF(rc.OIDCProviderCallback))
+	s.Router.GET(rc.configuration.LogoutURL, gin.WrapF(rc.Logout))
+}
+
+// AuthenticationSession is responsible for knowing if the user already has a valid credential or not.
+// If so, forward 200 OK + UserID Headers.
+// If not, begin OIDC Flow.
+func (rc RESTController) AuthenticationSession(w http.ResponseWriter, r *http.Request) {
+	// workflow to identify if there is a token present.
+	token, err := rc.getCredentials(r)
+	// if a token is not found, the OIDC flow will be started.
+	if authErrors.IsTokenNotFound(err) {
+		rc.RedirectToOIDCProvider(w, r)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+
+	// if an unexpected error occurs, the flow must be stopped.
+	if err != nil {
+		rc.logger.Error(err)
+		rc.showErrorResponse(w, err, http.StatusInternalServerError)
+		return
+	}
+
+	claims, err := rc.oidcService.Decode(token)
+	// check if the token is expired.
+	// if so, the OIDC flow will be started to recreate a token.
+	if authErrors.IsTokenExpired(err) {
+		rc.logger.Warn(err)
+		rc.RedirectToOIDCProvider(w, r)
+		return
+	}
+	// if a token was passed but it is not valid, the flow must be stopped.
+	if err != nil {
+		rc.logger.Errorf("an error occurred while decoding token: %v", err)
+		rc.showErrorResponse(w, err, http.StatusBadRequest)
+		return
+	}
+
+	// if the token passed is valid, let's get the TokenInfo to write in the header.
+	userInfo := rc.oidcService.GetUserInfo(claims)
+	rc.logger.Infof("Authorizing request for UserID: %v", userInfo.UID)
+
+	// set UserID Header + 200 OK
+	w.Header().Set(rc.configuration.UserIDHeader, userInfo.UID)
+	w.WriteHeader(http.StatusOK)
 }
 
 // RedirectToOIDCProvider is the handler responsible for redirecting to the OIDC Provider.
 func (rc RESTController) RedirectToOIDCProvider(w http.ResponseWriter, r *http.Request) {
+	rc.logger.Info("Starting OIDC flow")
 	oidcProviderURL, err := rc.oidcService.GetOIDCProviderURL()
 	if err != nil {
+		rc.logger.Errorf(err.Error())
```

Comment:
Why are you using `Errorf`? There is no formatting here.

Reply:
agree